libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/TripSessionLocationEngine.kt

```diff
@@ -24,55 +24,60 @@ import java.util.concurrent.TimeUnit
  * trip session is not using replay, use the [NavigationOptions.locationEngine].
  */
 internal class TripSessionLocationEngine constructor(
-    val navigationOptions: NavigationOptions
+    private val navigationOptions: NavigationOptions,
+    private val replayLocationEngineProvider: (MapboxReplayer) -> LocationEngine = {
+        ReplayLocationEngine(it)
+    }
 ) {
 
     val mapboxReplayer: MapboxReplayer by lazy { MapboxReplayer() }
 
-    private val replayLocationEngine: ReplayLocationEngine by lazy {
-        ReplayLocationEngine(mapboxReplayer)
+    private val replayLocationEngine: LocationEngine by lazy {
+        replayLocationEngineProvider.invoke(mapboxReplayer)
     }
-    private var locationEngine: LocationEngine = navigationOptions.locationEngine
+    private var activeLocationEngine: LocationEngine? = null
     private var onRawLocationUpdate: (Location) -> Unit = { }
 
+    private val locationEngineCallback = object : LocationEngineCallback<LocationEngineResult> {
+        override fun onSuccess(result: LocationEngineResult?) {
+            logD(LOG_CATEGORY) {
+                "successful location engine callback $result"
+            }
+            result?.locations?.lastOrNull()?.let {
+                logIfLocationIsNotFreshEnough(it)
+                onRawLocationUpdate(it)
+            }
+        }
+
+        override fun onFailure(exception: Exception) {
+            logD("location on failure exception=$exception", LOG_CATEGORY)
+        }
+    }
+
     @SuppressLint("MissingPermission")
     fun startLocationUpdates(isReplayEnabled: Boolean, onRawLocationUpdate: (Location) -> Unit) {
         logD(LOG_CATEGORY) {
             "starting location updates for ${if (isReplayEnabled) "replay " else ""}location engine"
         }
+        stopLocationUpdates()
         this.onRawLocationUpdate = onRawLocationUpdate
-        val locationEngine = if (isReplayEnabled) {
+        activeLocationEngine = if (isReplayEnabled) {
             replayLocationEngine
         } else {
             navigationOptions.locationEngine
         }
-        locationEngine.requestLocationUpdates(
+        activeLocationEngine?.requestLocationUpdates(
```

Comment:
`requestLocationUpdates` can be invoked multiple times with different callbacks and all of them will receive location updates.
So my question is: is stopping location updates here the right way to go? I mean what if we want to invoke `startLocationUpdates` twice with different `onRawLocationUpdate`? Right now the second invocation will cancel the first one.
By the way, having a single `locationEngineCallback` guarantees that there will be only one location update, but is this designed? Eve if we remove `stopLocationUpdates` from here the situation won't change: we'll still be getting raw location updates only in the second lambda.
I mean it's not clear from the API that that's the behaviour. So I'm asking whether it's how it's supposed to be.

Reply:
> is stopping location updates here the right way to go? I mean what if we want to invoke startLocationUpdates twice with different onRawLocationUpdate? Right now the second invocation will cancel the first one.

It is built to manage a single location engine subscription, because it would be sending duplicates if it had more than one subscription. The `activeLocationEngine` can change based on parameters, so it is unregistering to whatever the previous engine is.

There could be improvements to the logic, but i'm not seeing why it would be needed. Do you see an issue?

Reply:
I'm only saying that it's not obvious how to use this API. It looks like you can invoke `startLocationUpdates` twice with different callbacks, receive updates to both callbacks and then stop everything via  `stopLocationUpdates`.
As long as we use the API correctly, there is no issue.